src/data_provider.rs

```diff
@@ -1,174 +1,267 @@
+                                obj.update(result);
+                            }
+                        }
+                        Event::UnitRegistered(result) => {
+                            println!("Unit registered {:?}", result);
+                            store.append(&VMGObject::new(
+                                result.name,
+                                result.description,
+                                result.status,
+                                result.trust_level,
+                            ));
                         }
                     }
                 }
-                glib::ControlFlow::Continue
+                task_runner.take();
             });
         }
 
         fn fill_by_mock_data() -> ListStore {
             let init_store = ListStore::new::<VMGObject>();
             let mut vec: Vec<VMGObject> = Vec::new();
-            vec.push(VMGObject::new("VM1".to_string(), String::from("This is the file.pdf and very very long description"), 
-                    VMStatus::Running, TrustLevel::NotSecure));
-            vec.push(VMGObject::new("VM2".to_string(), String::from("Google Chrome"), VMStatus::Paused, TrustLevel::Secure));
-            vec.push(VMGObject::new("VM3".to_string(), String::from("AppFlowy"), VMStatus::Running, TrustLevel::Secure));
+            vec.push(VMGObject::new(
+                "VM1".to_string(),
+                String::from("This is the file.pdf and very very long description"),
+                VMStatus::Running,
+                TrustLevel::NotSecure,
+            ));
+            vec.push(VMGObject::new(
+                "VM2".to_string(),
+                String::from("Google Chrome"),
+                VMStatus::Paused,
+                TrustLevel::Secure,
+            ));
+            vec.push(VMGObject::new(
+                "VM3".to_string(),
+                String::from("AppFlowy"),
+                VMStatus::Running,
+                TrustLevel::Secure,
+            ));
             init_store.extend_from_slice(&vec);
-            return init_store;
+            init_store
         }
 
         pub fn get_store(&self) -> ListStore {
-            self.store.lock().unwrap().clone()
-        }
-
-        pub fn get_store_ref(&self) -> Arc<Mutex<ListStore>> {
-            self.store.clone()
+            self.store.deref().clone()
         }
 
         pub fn get_current_service_address(&self) -> (String, u16) {
-            self.service_address.clone().into_inner()
+            self.service_address.borrow().clone()
         }
 
-        pub fn set_service_address(&self, addr: String, port: u16) {//only when disconnected/watch stopped
-            if !(self.admin_client.try_write().is_err()) {
+        pub fn set_service_address(&self, addr: String, port: u16) {
+            //only when disconnected/watch stopped
+            if self.admin_client.try_write().is_ok() {
```

Comment:
Sometimes address and port are not updated, may be waiting for admin_client availability is better? We need to check it on ghaf.

Reply:
There is now the busyloop rebased from main branch; will check for a cleaner solution though, as keeping the UI thread busy isn't optimal.

Reply:
Made changes so that the `AdminClient` lives only in the worker thread. This allowed for simplifications, and the locking loop becomes totally unnecessary.